docs/reference/commandline/service_create.md

```diff
@@ -678,6 +678,41 @@ appends a new placement preference after all existing placement preferences.
 `--placement-pref-rm` removes an existing placement preference that matches the
 argument.
 
+### Specify memory requirements and constraints for a service (--reserve-memory and --limit-memory)
+
+If your service needs a minimum amount of memory in order to run correctly,
+you can use `--reserve-memory` to specify that the service should only be
+scheduled on a node with this much memory available to reserve. If no node is
+available that meets the criteria, the service is not scheduled, but remains in
+a pending state.
+
+The following example requires that 4GB of memory be available and reservable
+on a given node before scheduling the service to run on that node.
+
+```bash
+$ docker service create --reserve-memory=4GB --name=too-big nginx:alpine
+```
+
+The managers won't schedule a set of containers on a single node whose combined
+reservations exceed the memory available on that node.
+
+After a task is scheduled and running, `--reserve-memory` does not enforce a
+memory limit. Use `--limit-memory` to ensure that a task uses no more than a
+given amount of memory on a node. This example limits the amount of memory used
+by the task to 4GB. The task will be scheduled even if each of your nodes has
+only 2GB of memory, because `--limit-memory` is an upper limit.
+
+```bash
+$ docker service create --limit-memory=4GB --name=too-big nginx:alpine
+```
+
+Using `--reserve-memory` and `--limit-memory` does not guarantee that Docker
+will not use more memory on your host than you want. For instance, you could
+create many services, the sum of whose memory usage could exhaust the available
```

Comment:
In a way, you can prevent this, as long as you take into account that other (non-containerised) software is running on the host as well; if `--reserve-memory >= --limit-memory`, docker won't schedule a service on a host that doesn't have enough memory. `--limit-memory` will limit the service's memory to stay within that limit, so if every service has a memory-reservation and limit set, it won't saturate the host.

There _is_ a downside to doing this however; reserving memory also means that you may not make optimum use of the node. Take for instance a service that under normal circumstances uses 100MB of memory, but depending on load can "peak" at 500MB. _Reserving_ 500MB for that service (to guarantee it's able to have 500MB for those "peaks") results in 400MB of memory being wasted most of the time.

So, there's different approaches possible;

- Conservative; reserve 500MB, and limit to 500MB. Basically you're now treating the service/containers as VM's, and you may be throwing away a big advantage of container; a bigger density of services per host.
- "Flexible"; limit to 500MB (you never want the service to take up more than 500MB, because it should not require more, and if it does, something is wrong), and "reserve" something in between - more than 100MB, which is required for the service to do it's regular job, but less than 500MB, because it's only needed for shorter periods.

As mentioned earlier; the "flexible" approach allows you to get a higher density. More services can be packed on the same node, and other services can take advantage of the "idle time" and use the memory.

There's no "fixed rule", because these choices are really depending on your use case (can I take the risk that my service temporarily can't get enough memory?)